### PR TITLE
Add deployment controller

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
1. Finalize the workload plugin controller interface
2. Add the deployment controller
3. Add a deployment based demo
TODO: add  rollout planner UTs

### CODE REVIEW

File: docs/examples/deployment-rollout/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+```shell
+kubectl apply -f docs/examples/deployment-rollout/app-source-prep.yaml
+```
+Wait for the applicationConfiguration "test-rolling-v1" `Rolling Status` to be "RollingTemplated"
+
+4. Modify the application image and apply
+```shell
+kubectl apply -f docs/examples/deployment-rollout/app-target.yaml
+```
+Wait for the applicationConfiguration "test-rolling-v2" `Rolling Status` to be "RollingTemplated"
```
Comment (wonderflow) [inline docs/examples/deployment-rollout/README.md:27]: as we discussed,  these two steps need to be merged.

**Review by wonderflow — APPROVED** [review]
LGTM, tests need to be added

File: docs/examples/deployment-rollout/README.md (hunk 1 above)
Comment (ryanzhang-oss) [reply docs/examples/deployment-rollout/README.md:27]: step 3 and 4?

Comment (wonderflow) [reply docs/examples/deployment-rollout/README.md]: yes, right？

Comment (ryanzhang-oss) [reply docs/examples/deployment-rollout/README.md]: step 3 is to mark source is going to be rollout and step 4 is to modify the target. I am not sure how to combine these two. 

We can probably remove the step 5

**Pull request merged**